### DISCUSSION TITLE
Parameterizing Image Automation Marker

### DISCUSSION THREAD

**Discussion by wmiller112** [fluxcd/flux2#2742] [Q&A] [answered]
I have an app repo with kustomize directory in which I define a base with namespace, helmrelease, imagePolicy, and imageAutomationPolicy resources. I then have patches for different envs (`staging`/`prod`). These are deployed via kustomizations, in which I pass in a `branch=<branch-name>` var to get unique instances of all the resources mentioned above. 

I'm trying to understand how I can reference the imageAutomationPolicy because it doesn't seem like the marker comment supports variables. ie. 
`# {"$imagepolicy": "my-project-demo:my-project-demo"}` - works
`# {"$imagepolicy": "my-project-${branch}:my-project-${branch}"}` - doesn't work, when `branch=demo` is set via the kustomization and all other resources are created properly

I attempted to name the imageAutomationPolicies consistently for an app, but since the marker also requires namespace, this didn't work, and doing the reverse (putting all imageAutomationPolicies in flux-system) also wouldnt work because they would need unique names. It seems that without variable support here, theres no way to seamlessly deploy an app from a repo, as you'd always have to update the helmrelease to point to the unique imageAutomationPolicy.

**Comment by kingdonb** [comment, marked as answer]
You will not be able to parameterize those because the Kyaml setter is read directly from the file, and not through any string interpolation engines. If you want to make a solution to keep your configuration DRY, the two strategies which are promoted and recommended are (1) in-cluster policy managers like Kyverno can generate resources inside of the cluster, or (2) a CI process can write the results of a manifest generation (jsonnet, cue, etc) process back to git, where Flux can read it directly.

There is no in-process manifest generation in Flux other than what Kustomize provides with its patching, environment substition, etc.
